src/jsonpath/parser.c

```diff
@@ -356,6 +356,12 @@ static struct ast_node* parse_primary(PARSER_PARAMS) {
       }
       strcpy(tail->data.d_selector.value, CUR_TOKEN_LITERAL());
       CONSUME_TOKEN();
+
+      if (CUR_TOKEN() == LEX_WILD_CARD) {
+        free_ast_nodes(tail);
```

Comment:
It doesn't look like the `AST_SELECTOR`s are cleaned up here, maybe we should call `free_ast_nodes(ret);`?

Reply:
Hmm not seeing any issues in valgrind, not exactly sure why it's working.

Reply:
I can trigger a memory leak.

Update the child path to have 2 nodes:

```diff
-$result = $jsonPath->find($data, "$[?(@.key*2==100)]");
+$result = $jsonPath->find($data, "$[?(@.keyA.keyB*2==100)]");
```

Comment out the exception:
```diff
-zend_throw_exception(spl_ce_RuntimeException, "Multiplying node values is not supported.", 0);
```

Get this result:
```
[Tue May  4 15:15:51 2021]  Script:  '/home/mike/php-ext-jsonpath/tests/comparison_filter/053.php'
/home/mike/php-ext-jsonpath/src/jsonpath/parser.c(55) :  Freeing 0x00007f2566e7e150 (72 bytes), script=/home/mike/php-ext-jsonpath/tests/comparison_filter/053.php
=== Total 1 memory leaks detected ===
```

It looks like throwing an exception might automatically clean up any allocated memory... interesting.

Reply:
The leak is fixed with this:

```diff
-free_ast_nodes(tail);
+free_ast_nodes(ret);
```

It looks like we may not need to free allocations when an exception is thrown.. will look into this. Maybe for now just update the call to free the head of the list?

Reply:
Another way to reproduce the problem is to disable the Zend Memory manager, which uses the system's built-in `malloc` for memory allocation.

Test file:
```php
// ...
$result = $jsonPath->find($data, "$[?(@.keyA.keyB.keyC*2==100)]");
// ...
```

Disable Zend MM, observe leak:

```bash
mike@mike-VirtualBox:~/php-ext-jsonpath$ export USE_ZEND_ALLOC=0
mike@mike-VirtualBox:~/php-ext-jsonpath$ valgrind php test.php
==27322== Memcheck, a memory error detector
==27322== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==27322== Using Valgrind-3.13.0 and LibVEX; rerun with -h for copyright info
==27322== Command: php test.php
==27322== 

Fatal error: Uncaught RuntimeException: Multiplying node values is not supported. in /home/mike/php-ext-jsonpath/test.php:22
Stack trace:
#0 /home/mike/php-ext-jsonpath/test.php(22): JsonPath->find(Array, '$[?(@.keyA.keyB...')
#1 {main}
  thrown in /home/mike/php-ext-jsonpath/test.php on line 22
==27322== 
==27322== HEAP SUMMARY:
==27322==     in use at exit: 8,806 bytes in 49 blocks
==27322==   total heap usage: 11,444 allocs, 11,395 frees, 2,163,491 bytes allocated
==27322== 
==27322== LEAK SUMMARY:
==27322==    definitely lost: 448 bytes in 3 blocks
==27322==    indirectly lost: 4,280 bytes in 7 blocks
==27322==      possibly lost: 2,952 bytes in 18 blocks
==27322==    still reachable: 1,126 bytes in 21 blocks
==27322==         suppressed: 0 bytes in 0 blocks
==27322== Rerun with --leak-check=full to see details of leaked memory
==27322== 
==27322== For counts of detected and suppressed errors, rerun with: -v
==27322== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

Disable Zend MM, no leak:

```bash
mike@mike-VirtualBox:~/php-ext-jsonpath$ export USE_ZEND_ALLOC=1
mike@mike-VirtualBox:~/php-ext-jsonpath$ valgrind php test.php
==27326== Memcheck, a memory error detector
==27326== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==27326== Using Valgrind-3.13.0 and LibVEX; rerun with -h for copyright info
==27326== Command: php test.php
==27326== 

Fatal error: Uncaught RuntimeException: Multiplying node values is not supported. in /home/mike/php-ext-jsonpath/test.php:22
Stack trace:
#0 /home/mike/php-ext-jsonpath/test.php(22): JsonPath->find(Array, '$[?(@.keyA.keyB...')
#1 {main}
  thrown in /home/mike/php-ext-jsonpath/test.php on line 22
==27326== 
==27326== HEAP SUMMARY:
==27326==     in use at exit: 1,126 bytes in 21 blocks
==27326==   total heap usage: 10,998 allocs, 10,977 frees, 1,694,266 bytes allocated
==27326== 
==27326== LEAK SUMMARY:
==27326==    definitely lost: 0 bytes in 0 blocks
==27326==    indirectly lost: 0 bytes in 0 blocks
==27326==      possibly lost: 0 bytes in 0 blocks
==27326==    still reachable: 1,126 bytes in 21 blocks
==27326==         suppressed: 0 bytes in 0 blocks
==27326== Rerun with --leak-check=full to see details of leaked memory
==27326== 
==27326== For counts of detected and suppressed errors, rerun with: -v
==27326== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

Reply:
Cool, thanks. I changed `tail` to `ret`.

We'll need to add freeing in a few other places as well. :slightly_smiling_face: